signal/playerctl/playerctl_lib.lua

```diff
@@ -95,8 +96,9 @@ local function metadata_cb(player, metadata)
     if player == manager.players[1] then
         -- Callback can be called even though values we care about haven't
         -- changed, so check to see if they have
-        if player ~= last_player or title ~= last_title or
-           artist ~= last_artist or artUrl ~= last_artUrl
+        index = index + 1
+        if (player ~= last_player or title ~= last_title or
+           artist ~= last_artist) and (artUrl ~= "" or index == 2)
```

Comment:
> I think part of this issue can and should be handled by client side code rather than the server side code here. If receiving an empty string for the artUrl is an issue, that's a case that should be handled by client side code because as you say some media players don't provide an artUrl but those signals are still valid

The issue I had is that when artUrl was empty, when passed to the signal it would still somehow point to the previous artUrl that did have image, causing the notification spawned by the signal to show the last media artUrl. 

> As for the case where the title or artist changes but the artUrl does not, we could emit an empty string instead of downloading the same image and sending a duplicate artUrl. 
The artUrl does change on the first message, it's just always ends up being empty. The second signal only triggers because on the first message artUrl was empty so artUrl ~= last_artUrl becomes true
That media might still have an artUrl, but you can't know that at the first message. 

If you pass an empty string, you will still have duplicate message

> There's a lot of variation on how different applications handle MPRIS so trying to compensate for some of them that produce multiple signals and other that don't leads down a bit of a rabbit hole. Let me know what you think @JavaCafe01.

I did test it with few apps, but obviously not all of them. Maybe only use this method for what we already know is working?

Reply:
So just so I understand, you see a signal where the title and artist change but the artUrl is still the same as the last signal? Other times you get a signal where the artUrl field is empty followed by another that has is correctly filled right? I think the changes you've made work for apps that exhibit this behavior but would also reject some legitimate signals from apps with different behavior.  If these double signals come in quick succession I'm wondering if a form of debouncing would be applicable where we simply wait a small amount of time before emitting the signal. If another signal comes in that time the info can be updated before actually emitting. That way there would only be a single signal emitted with (hopefully) the right info. That depends on the timing between the signals though.

Reply:
Here's the dbus-monitor output:
```
string "org.mpris.MediaPlayer2.Player"
   array [
      dict entry(
         string "Metadata"
         variant             array [
               dict entry(
                  string "mpris:trackid"
                  variant                      object path "/org/mpris/MediaPlayer2/firefox"
               )
               dict entry(
                  string "xesam:title"
                  variant                      string "HD Video (1080p) with Relaxing Music of Native American Shamans"
               )
               dict entry(
                  string "xesam:album"
                  variant                      string ""
               )
               dict entry(
                  string "xesam:artist"
                  variant                      array [
                        string "LoungeV Films - Relaxing Music and Nature Sounds"
                     ]
               )
            ]
      )
   ]
   array [
   ]
   string "org.mpris.MediaPlayer2.Player"
   array [
      dict entry(
         string "Metadata"
         variant             array [
               dict entry(
                  string "mpris:trackid"
                  variant                      object path "/org/mpris/MediaPlayer2/firefox"
               )
               dict entry(
                  string "xesam:title"
                  variant                      string "HD Video (1080p) with Relaxing Music of Native American Shamans"
               )
               dict entry(
                  string "xesam:album"
                  variant                      string ""
               )
               dict entry(
                  string "xesam:artist"
                  variant                      array [
                        string "LoungeV Films - Relaxing Music and Nature Sounds"
                     ]
               )
            ]
      )
   ]
```

As you can see, the first message doesn't provide the artUrl, resulting in it being empty in Awesome while the second message does contain the artUrl. 
Considering this if statement:
```
        if player ~= last_player or title ~= last_title or
           artist ~= last_artist or artUrl ~= last_artUrl
```
This check alone should have been enough to prevent those double signals, but since artUrl was empty the first time, and not on the 2nd, this statement becomes true and emits a new signal. That's issue #1 
Issue #2 is  that somehow when that artUrl is empty, the signals are getting emitted with the last played media that had the artUrl field. That's why I added a new signal without passing in the artUrl.

Reply:
@Kasper24 I actually think your original changes may be closer to a better fix for the issue where the last song's album art is emitted with updated title and artist. I think we just need to have a check for artUrl being empty and if so emit the signal without attempting to download anything. That way the client side code will receive an empty string for the artUrl which will indicate none is available. We will save the overhead of launching a shell and running those shell commands when we know there is no URL to download from. I also think an empty string as the artUrl parameter is a better indicator to client code that there is no album art than passing a path to a file that has been removed. I'm thinking something list this which is similar to your original changes

```lua
if artUrl ~= "" and artUrl ~= last_artUrl then
    awful.spawn.with_line_callback(get_album_art(artUrl), {
        stdout = function(line)
            awesome.emit_signal(
                "bling::playerctl::title_artist_album",
                title,
                artist,
                line,
                player.player_name
            )
        end
    })
else
    awesome.emit_signal(
        "bling::playerctl::title_artist_album",
        title,
        artist,
        "",
        player.player_name
    )
end
```